x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/indices/recovery/SearchableSnapshotRecoveryState.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.indices.recovery;
+
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.cluster.routing.ShardRouting;
+import org.elasticsearch.common.Nullable;
+
+import java.util.HashSet;
+import java.util.Set;
+
+public final class SearchableSnapshotRecoveryState extends RecoveryState {
+    private boolean preWarmFinished;
+
+    public SearchableSnapshotRecoveryState(ShardRouting shardRouting, DiscoveryNode targetNode, @Nullable DiscoveryNode sourceNode) {
+        super(shardRouting, targetNode, sourceNode, new Index());
+    }
+
+    @Override
+    public synchronized RecoveryState setStage(Stage stage) {
+        // The transition to the final state was done by #prewarmCompleted, just ignore the transition
+        if (getStage() == Stage.DONE) {
+            return this;
+        }
+
+        // Pre-warm is still running, hold the state transition
+        // until the pre-warm process finishes
+        if (preWarmFinished == false && stage == Stage.DONE) {
+            return this;
```

Comment:
should we throw an exception here if the stage is currently not `FINALIZE`, i.e. `validateCurrentStage(FINALIZE)`?

Reply:
+1